src/Raven.Client/Documents/Identity/AsyncHiLoIdGenerator.cs

```diff
@@ -83,25 +83,37 @@ public async Task<long> NextIdAsync()
         {
             while (true)
             {
-                //local range is not exhausted yet
+                var current = _nextRangeTask;
+
+                // local range is not exhausted yet
                 var range = Range;
                 var id = Interlocked.Increment(ref range.Current);
                 if (id <= range.Max)
                     return id;
 
-                //local range is exhausted , need to get a new range
-                var maybeNextTask = new Lazy<Task>(GetNextRangeAsync);
-
-                var nextTask = Interlocked.CompareExchange(ref _nextRangeTask,
-                                   maybeNextTask, null) ?? maybeNextTask;
                 try
                 {
-                    await nextTask.Value.ConfigureAwait(false);
+                    // let's try to call the existing task for next range
+                    await current.Value.ConfigureAwait(false);
+                    if (range != Range)
+                        continue;
+                }
+                catch
+                {
+                    // previous task was faulted, we will replace it
                 }
-                finally
+
+                // local range is exhausted , need to get a new range
+                var maybeNextTask = new Lazy<Task>(GetNextRangeAsync);
+                var nextTask = Interlocked.CompareExchange(ref _nextRangeTask, maybeNextTask, current);
+                if (nextTask == current) // replace was successful
                 {
-                    Interlocked.CompareExchange(ref _nextRangeTask, null, nextTask);
+                    await maybeNextTask.Value.ConfigureAwait(false);
+                    continue;
                 }
+
+                // failed to replace, let's wait on the previous task
+                await nextTask.Value.ConfigureAwait(false);
```

Comment:
This should be wrapped in try/catch and ignored.
Someone else failed, but we'll retry until our own task failed

Reply:
done